### PR TITLE
feat: add request header in match context

### PR DESCRIPTION
**Pull request by msumit**
## Related issue
https://github.com/ory/oathkeeper/issues/512

## Proposed changes
After wasting a lot of time on why I'm not able to use Headers within the Remote Authorizer, I found out that Oathkeeper isn't setting the Headers while initializing the AuthenticationSession. We use Remote Authorizer and pass on some key information through headers only & without this fix, we can't use Oathkeeper at all, so setting the request headers inside MatchContext. 

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security. vulnerability, I
      confirm that I got green light (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [x] I have added tests that prove my fix is effective or that my feature
      works.
- [ ] I have added or changed [the documentation](docs/docs).

## Further comments
See this PR -> https://github.com/ory/oathkeeper/pull/718

### CODE REVIEW

File: proxy/request_handler_test.go
Comment (andriisoldatenko) [inline proxy/request_handler_test.go:47]:
```
return &http.Request{URL: x.ParseURLOrPanic(u), Method: "GET", Header: http.Header{"Test-Header":[]string{"Test-Value"}}}
```

File: proxy/request_handler_test.go
Comment (andriisoldatenko) [inline proxy/request_handler_test.go:499]: i think for expected results better to use constant / or literal instead of function

**Comment by msumit** [comment]
@aeneasr can you plz review.

**Review by aeneasr — CHANGES_REQUESTED** [review]
Thank you this looks great and sorry for the late review - last week was hectic :)

Could you please also update the documentation in `docs/docs` on this? Thank you!

**Comment by aeneasr** [comment]
Thank you, this looks great! The CI is failing because some files are formatted incorrectly. To format them, run:

```
$ make format
$ git commit -a -m "styles: format code"
$ git push
```

Thank you!

**Comment by aeneasr** [comment]
Hmmm...why isn't the pipeline triggering -.-

**Comment by aeneasr** [comment]
Yeah it's not your fault!

**Comment by aeneasr** [comment]
But `ci/circleci: test` should be running and it's not :/ I have one guide regarding this:

---

Unfortunately, for some reason, the CircleCI tests are not running. Do you maybe follow your/a fork of this repository on CircleCI? If so, you need to unsubscribe / unwatch from that CircleCI project. Then, make another push to your branch using:

```
$ git commit --amend --allow-empty
$ git push --force
```

That should get the CI running! Thank you :)

**Comment by msumit** [comment]
@aeneasr I think you need to approve something somewhere to kick off the tests. 
`First-time contributors need a maintainer to approve running workflows. Learn more.`

**Comment by msumit** [comment]
@aeneasr can you look into this

**Comment by aeneasr** [comment]
Hey there, sorry for the long delay - yesterday was a holiday :) I've approved the run. [I've also create a post on GH community](https://github.community/t/approve-run-is-a-major-pain-when-managing-large-oss-ecosystems/179765) regarding this as the current approval behavior is so annoying.

**Comment by aeneasr** [comment]
Hm no, this looks like a circle ci issue. I'll run the tests locally

**Review by aeneasr — APPROVED** [review]

**Comment by aeneasr** [comment]
Awesome, thank you! 🎉 Your contribution makes Ory better :)

**Pull request merged**
